### PR TITLE
Allow filters with same name and different number of parameters from different classes

### PR DESCRIPTION
**Pull request by BenJBailey**
Prior to this change, if a class implemented a filter with the same name as another filter method in another class, yet the filter had a different number of parameters, that filter would be ignored.

So say that a class has the filter `| DoThing: 1`. and another class has a filter `| DoThing`. If the methods were added from the first filter's class before the second filter's class was added, only `| DoThing` would work. Trying to use `| DoThing: 1` would result in an error.

This also allows the extension of DotLIquid's StandardFilters.

- Changed `Strainer.Extend()` to check the name and number of parameters for existing filters before removing methods from `_methods`.
- Refactored check for count of non Context method parameters to extension method.
- Added test case for different methods with the same name and number of parameters in different classes.
- Added test case for different methods with the same name and number of parameters in different classes with the context parameter.
- Added test case for overriding methods that have the same name and the same number of parameters with the last class to be added to Filters.

### CODE REVIEW

File: src/DotLiquid.Tests/FilterTests.cs [hunk 1]
```diff
@@ -83,6 +83,56 @@ public static string ConcatWithContext(Context context, string one, string two,
...
+
+        private static class FiltersWithMulitpleMethodSignaturesDifferentClassesWithContextParamTwo
+        {
+            public static string ConcatWithContext(Context context, string one, string two, string three)
+            {
+                return string.Concat(one, two, three);
+            }
+        }
+
+        private static class FiltersWithMulitpleMethodSignaturesDifferentClassesWithContextParamOne
```
Comment (microalps) [inline src/DotLiquid.Tests/FilterTests.cs:127]: Nit: spelling

File: src/DotLiquid.Tests/FilterTests.cs [hunk 2]
```diff
@@ -83,6 +83,56 @@ public static string ConcatWithContext(Context context, string one, string two,
             }
         }
 
+        private static class FiltersWithMulitpleMethodSignaturesDifferentClassesOne
```
Comment (microalps) [inline src/DotLiquid.Tests/FilterTests.cs:86]: Nit: spelling

File: src/DotLiquid.Tests/FilterTests.cs [hunk 3]
```diff
@@ -206,6 +256,36 @@ public void TestFilterWithMultipleMethodSignaturesAndContextParam()
             Assert.AreEqual("ABC", Template.Parse("{{'A' | concat_with_context : 'B', 'C'}}").Render());
         }
 
+        [Test]
+        public void TestFilterWithMultipleMethodSignaturesDifferentClasses()
+        {
+            Template.RegisterFilter(typeof(FiltersWithMulitpleMethodSignaturesDifferentClassesOne));
```
Comment (microalps) [inline src/DotLiquid.Tests/FilterTests.cs:262]: Since this is global, the behavior is sort of unpredictable across tests. Should we make a test of filters local to the context.

**Review by microalps — COMMENTED** [review]
Seems like a simple fix. Might need better testing for context level filters

File: src/DotLiquid.Tests/FilterTests.cs (hunk 2 above)
Comment (BenJBailey) [reply src/DotLiquid.Tests/FilterTests.cs]: Resolved here : https://github.com/dotliquid/dotliquid/pull/453/commits/e9eccc68a98ca7865bdd7ad3dcd1c23761dc8b5e

File: src/DotLiquid.Tests/FilterTests.cs (hunk 1 above)
Comment (BenJBailey) [reply src/DotLiquid.Tests/FilterTests.cs]: Resolved here: https://github.com/dotliquid/dotliquid/pull/453/commits/e9eccc68a98ca7865bdd7ad3dcd1c23761dc8b5e

File: src/DotLiquid.Tests/FilterTests.cs (hunk 3 above)
Comment (BenJBailey) [reply src/DotLiquid.Tests/FilterTests.cs]: Added context tests here: https://github.com/dotliquid/dotliquid/pull/453/commits/67a6be4bf74ce0ec2a7b86c6110f6d2b9d069523

Comment (microalps) [reply src/DotLiquid.Tests/FilterTests.cs]: I was referring to localFilters attribute as used [here](https://github.com/dotliquid/dotliquid/blob/master/src/DotLiquid.Tests/ShopifyFiltersTests.cs)

Comment (BenJBailey) [reply src/DotLiquid.Tests/FilterTests.cs]: Ah, okay! I added some tests to cover that case here: https://github.com/dotliquid/dotliquid/pull/453/commits/31139cdc569c2a28e6ecdc63b36db3f80a7fd12e

File: src/DotLiquid.Tests/ConditionTests.cs [hunk 4]
```diff
@@ -657,7 +657,7 @@ public void TestRuby_PascalCaseNotAccepted()
         public void TestCSharp_LowerCaseAccepted()
         {
             Helper.AssertTemplateResult("", "{% if 'bob' startswith 'B' %} YES {% endif %}", null, new CSharpNamingConvention());
-            Helper.AssertTemplateResult(" YES ", "{% if 'Bob' startswith 'B' %} YES {% endif %}", null, new CSharpNamingConvention());
+            Helper. AssertTemplateResult(" YES ", "{% if 'Bob' startswith 'B' %} YES {% endif %}", null, new CSharpNamingConvention());
```
Comment (daviburg) [inline src/DotLiquid.Tests/ConditionTests.cs:660]: Please revert this change. Space between a class name and method is atypical coding style.

File: src/DotLiquid.Tests/FilterTests.cs [hunk 5]
```diff
@@ -83,6 +83,56 @@ public static string ConcatWithContext(Context context, string one, string two,
...
+            }
+        }
+
+        private static class FiltersWithMultipleMethodSignaturesDifferentClassesWithContextParamOne
+        {
+            public static string ConcatWithContext(Context context, string one, string two)
+            {
+                return string.Concat(one, two);
+            }
+
```
Comment (daviburg) [inline src/DotLiquid.Tests/FilterTests.cs:133]: nit
```suggestion
```

File: src/DotLiquid.Tests/FilterTests.cs [hunk 6]
```diff
@@ -188,15 +238,26 @@ public void TestFilterWithTwoArgumentsLong()
             Assert.AreEqual("[1150]", new Variable("var | add_sub: 200, 50").Render(_context));
         }
 
+
```
Comment (daviburg) [inline src/DotLiquid.Tests/FilterTests.cs:241]: nit
```suggestion
```

File: src/DotLiquid.Tests/FilterTests.cs [hunk 7]
```diff
@@ -188,15 +238,26 @@ public void TestFilterWithTwoArgumentsLong()
...
         public void TestFilterWithMultipleMethodSignatures()
         {
-            Template.RegisterFilter(typeof(FiltersWithMulitpleMethodSignatures));
+            Template.RegisterFilter(typeof(FiltersWithMultipleMethodSignatures));
 
             Assert.AreEqual("AB", Template.Parse("{{'A' | concatenate : 'B'}}").Render());
             Assert.AreEqual("ABC", Template.Parse("{{'A' | concatenate : 'B', 'C'}}").Render());
         }
 
+
```
Comment (daviburg) [inline src/DotLiquid.Tests/FilterTests.cs:251]: always a single white line
```suggestion
```

File: src/DotLiquid.Tests/FilterTests.cs [hunk 8]
```diff
@@ -206,6 +267,111 @@ public void TestFilterWithMultipleMethodSignaturesAndContextParam()
...
+            _context.AddFilters(typeof(FiltersWithMultipleMethodSignaturesDifferentClassesTwo));
+
+            Assert.AreEqual("AB", new Variable("'A' | concatenate : 'B'").Render(_context));
+            Assert.AreEqual("ABC", new Variable("'A' | concatenate : 'B', 'C'").Render(_context));
+        }
+
+        [Test]
+        public void TestFilterAsLocalFilterWithMultipleMethodSignaturesDifferentClasses()
+        {
+
```
Comment (daviburg) [inline src/DotLiquid.Tests/FilterTests.cs:302]:
```suggestion
```

File: src/DotLiquid.Tests/FilterTests.cs [hunk 9]
```diff
@@ -206,6 +267,111 @@ public void TestFilterWithMultipleMethodSignaturesAndContextParam()
...
+            Helper.AssertTemplateResult(
+                expected: "AB // ABC",
+                template: "{{'A' | concatenate : 'B'}} // {{'A' | concatenate : 'B', 'C'}}",
+                localVariables: null,
+                localFilters: new[] { typeof(FiltersWithMultipleMethodSignaturesDifferentClassesOne), typeof(FiltersWithMultipleMethodSignaturesDifferentClassesTwo) });
+        }
+
+
```
Comment (daviburg) [inline src/DotLiquid.Tests/FilterTests.cs:310]:
```suggestion
```

File: src/DotLiquid.Tests/FilterTests.cs [hunk 10]
```diff
@@ -206,6 +267,111 @@ public void TestFilterWithMultipleMethodSignaturesAndContextParam()
...
+        public void TestFilterInContextOverridesMethodWithSameMethodSignaturesDifferentClasses()
+        {
+            _context.AddFilters(typeof(FilterWithSameMethodSignatureDifferentClassOne));
+            _context.AddFilters(typeof(FilterWithSameMethodSignatureDifferentClassTwo));
+
+            Assert.AreEqual("ABClass Two", new Variable("'A' | concatenate : 'B'").Render(_context));
+            Assert.AreNotEqual("ABClass One", new Variable("'A' | concatenate : 'B'").Render(_context));
+        }
+
+
```
Comment (daviburg) [inline src/DotLiquid.Tests/FilterTests.cs:363]:
```suggestion
```

File: src/DotLiquid.Tests/FilterTests.cs [hunk 11]
```diff
@@ -206,6 +267,111 @@ public void TestFilterWithMultipleMethodSignaturesAndContextParam()
...
+            Assert.AreEqual("ABC", new Variable("'A' | concat_with_context : 'B', 'C'").Render(_context));
+        }
+
+        [Test]
+        public void TestFilterOverridesMethodWithSameMethodSignaturesDifferentClasses()
+        {
+            Template.RegisterFilter(typeof(FilterWithSameMethodSignatureDifferentClassOne));
+            Template.RegisterFilter(typeof(FilterWithSameMethodSignatureDifferentClassTwo));
+
+            Assert.AreEqual("ABClass Two", Template.Parse("{{'A' | concatenate : 'B'}}").Render());
```
Comment (daviburg) [inline src/DotLiquid.Tests/FilterTests.cs:349]: Please add a code comment explaining why this order of preference is happening when the method signatures are the same.

File: src/DotLiquid.Tests/FilterTests.cs [hunk 12]
```diff
@@ -206,6 +267,111 @@ public void TestFilterWithMultipleMethodSignaturesAndContextParam()
...
+        {
+
+            Helper.AssertTemplateResult(
+                expected: "AB // ABC",
+                template: "{{'A' | concat_with_context : 'B'}} // {{'A' | concat_with_context : 'B', 'C'}}",
+                localVariables: null,
+                localFilters: new[] { typeof(FiltersWithMultipleMethodSignaturesDifferentClassesWithContextParamOne), typeof(FiltersWithMultipleMethodSignaturesDifferentClassesWithContextParamTwo) });
+        }
+
+
```
Comment (daviburg) [inline src/DotLiquid.Tests/FilterTests.cs:332]:
```suggestion
```

File: src/DotLiquid.Tests/FilterTests.cs [hunk 13]
```diff
@@ -206,6 +267,111 @@ public void TestFilterWithMultipleMethodSignaturesAndContextParam()
...
+        public void TestFilterAsLocalOverridesMethodWithSameMethodSignaturesDifferentClasses()
+        {
+            Helper.AssertTemplateResult(
+                           expected: "ABClass Two",
+                           template: "{{'A' | concatenate : 'B'}}",
+                           localVariables: null,
+                           localFilters: new[] { typeof(FilterWithSameMethodSignatureDifferentClassOne), typeof(FilterWithSameMethodSignatureDifferentClassTwo) });
+        }
+
+
```
Comment (daviburg) [inline src/DotLiquid.Tests/FilterTests.cs:374]:
```suggestion
```

File: src/DotLiquid/Strainer.cs [hunk 14]
```diff
@@ -15,6 +15,27 @@ static class DictionaryExtensions
                 return dic[key] = factory();
             return found;
         }
+
```
Comment (daviburg) [inline src/DotLiquid/Strainer.cs:18]:
```suggestion
```

File: src/DotLiquid/Strainer.cs [hunk 15]
```diff
@@ -15,6 +15,27 @@ static class DictionaryExtensions
                 return dic[key] = factory();
             return found;
         }
+
+    }
+
+    static class MethodInfoExtensions
```
Comment (daviburg) [inline src/DotLiquid/Strainer.cs:21]: specify the access explicitly always (public/protected/internal/private)

File: src/DotLiquid/Strainer.cs [hunk 16]
```diff
@@ -15,6 +15,27 @@ static class DictionaryExtensions
                 return dic[key] = factory();
             return found;
         }
+
+    }
+
+    static class MethodInfoExtensions
+    {
+        public static int GetNonContextParameterCount(this MethodInfo method)
```
Comment (daviburg) [inline src/DotLiquid/Strainer.cs:23]: Add XML comments for public classes and public methods.

File: src/DotLiquid/Strainer.cs [hunk 17]
```diff
@@ -15,6 +15,27 @@ static class DictionaryExtensions
...
             return found;
         }
+
+    }
+
+    static class MethodInfoExtensions
+    {
+        public static int GetNonContextParameterCount(this MethodInfo method)
+        {
+            return method.GetParameters().Count(p => p.ParameterType != typeof(Context));
```
Comment (daviburg) [inline src/DotLiquid/Strainer.cs:25]: avoid single letter variable names for clarity. Always use complete words from English dictionary, unless there is a well-accepted acronym.

File: src/DotLiquid/Strainer.cs [hunk 18]
```diff
@@ -15,6 +15,27 @@ static class DictionaryExtensions
...
+        }
+
+        public static bool MatchesMethod(this MethodInfo method, KeyValuePair<string, IList<Tuple<object, MethodInfo>>> compareMethod)
+        {
+            string methodName = Template.NamingConvention.GetMemberName(method.Name);
+            if (compareMethod.Key != methodName)
+            {
+                return false;
+            }
+            int methodParamCount = method.GetNonContextParameterCount();
```
Comment (daviburg) [inline src/DotLiquid/Strainer.cs:35]:
```suggestion

            var methodParamCount = method.GetNonContextParameterCount();
```

File: src/DotLiquid/Strainer.cs [hunk 19]
```diff
@@ -15,6 +15,27 @@ static class DictionaryExtensions
...
+    static class MethodInfoExtensions
+    {
+        public static int GetNonContextParameterCount(this MethodInfo method)
+        {
+            return method.GetParameters().Count(p => p.ParameterType != typeof(Context));
+        }
+
+        public static bool MatchesMethod(this MethodInfo method, KeyValuePair<string, IList<Tuple<object, MethodInfo>>> compareMethod)
+        {
+            string methodName = Template.NamingConvention.GetMemberName(method.Name);
```
Comment (daviburg) [inline src/DotLiquid/Strainer.cs:30]: inline single use parameters

File: src/DotLiquid/Strainer.cs [hunk 20]
```diff
@@ -15,6 +15,27 @@ static class DictionaryExtensions
...
+            if (compareMethod.Key != methodName)
+            {
+                return false;
+            }
+            int methodParamCount = method.GetNonContextParameterCount();
+
+            return compareMethod.Value.Any(m => m.Item2.GetNonContextParameterCount() == methodParamCount);
```
Comment (daviburg) [inline src/DotLiquid/Strainer.cs:37]: should you also be checking if the parameter types are matching, not just their count?

File: src/DotLiquid/Strainer.cs [hunk 21]
```diff
@@ -66,26 +87,34 @@ public Strainer(Context context)
             _context = context;
         }
 
+
```
Comment (daviburg) [inline src/DotLiquid/Strainer.cs:90]:
```suggestion
```

File: src/DotLiquid/Strainer.cs [hunk 22]
```diff
@@ -66,26 +87,34 @@ public Strainer(Context context)
...
         /// <summary>
         /// In this C# implementation, we can't use mixins. So we grab all the static
         /// methods from the specified type and use them instead.
         /// </summary>
         /// <param name="type"></param>
         public void Extend(Type type)
         {
-            // From what I can tell, calls to Extend should replace existing filters. So be it.
+            // From what I can tell, calls to Extend should replace existing filters with the same number of params. So be it.
```
Comment (daviburg) [inline src/DotLiquid/Strainer.cs:98]: Probably a good time to revise the wording of this comment. "From what I can tell," and "So be it." does not add value but adds doubt in the implementation. Remove it assuming we are now asserting this is the behavior we specify.

File: src/DotLiquid/Strainer.cs [hunk 23]
```diff
@@ -66,26 +87,34 @@ public Strainer(Context context)
...
+                }
+            }
 
             foreach (MethodInfo methodInfo in methods)
             {
                 AddMethodInfo(methodInfo.Name, null, methodInfo);
             } // foreach
         }
 
+
```
Comment (daviburg) [inline src/DotLiquid/Strainer.cs:115]: remove all these duplicate white lines

File: src/DotLiquid.Tests/FilterTests.cs [hunk 24]
```diff
@@ -206,6 +267,111 @@ public void TestFilterWithMultipleMethodSignaturesAndContextParam()
...
+        }
+
+        [Test]
+        public void TestFilterInContextOverridesMethodWithSameMethodSignaturesDifferentClasses()
+        {
+            _context.AddFilters(typeof(FilterWithSameMethodSignatureDifferentClassOne));
+            _context.AddFilters(typeof(FilterWithSameMethodSignatureDifferentClassTwo));
+
+            Assert.AreEqual("ABClass Two", new Variable("'A' | concatenate : 'B'").Render(_context));
+            Assert.AreNotEqual("ABClass One", new Variable("'A' | concatenate : 'B'").Render(_context));
```
Comment (daviburg) [inline src/DotLiquid.Tests/FilterTests.cs:360]: It would be wise to have at least one test case which ends up calling class one so we are not picking class two for another unintended reason that filter add order.

**Review by daviburg — CHANGES_REQUESTED** [review]

**Comment by daviburg** [comment]
> Looks good to me @daviburg please review.

I'm surprised you didn't see me coming on the coding style. :-D

File: src/DotLiquid/Strainer.cs (hunk 17 above)
Comment (daviburg) [reply src/DotLiquid/Strainer.cs:25]: And yes we do have existing code that violates this, not particularly easy to decypher.

File: src/DotLiquid/Strainer.cs (hunk 16 above)
Comment (daviburg) [reply src/DotLiquid/Strainer.cs:23]: And yes we do have existing code that violates this, not particularly easy to decipher.

File: src/DotLiquid/Strainer.cs (hunk 20 above)
Comment (microalps) [reply src/DotLiquid/Strainer.cs:37]: @daviburg We historically didn't do type matching. Liquid doesn't really care and most of the time we accept object. It would be a good enhancement as we can optimize lots of code that does casting unnecessarily, but it wouldn't be in scope for this.

Comment (daviburg) [reply src/DotLiquid/Strainer.cs:37]: fair enough

File: src/DotLiquid.Tests/ConditionTests.cs (hunk 4 above)
Comment (BenJBailey) [reply src/DotLiquid.Tests/ConditionTests.cs]: Fixed: https://github.com/dotliquid/dotliquid/pull/453/commits/42ea12d8664736c8ab5a3aeff8f9f62e3049fdd2

File: src/DotLiquid.Tests/FilterTests.cs (hunk 11 above)
Comment (BenJBailey) [reply src/DotLiquid.Tests/FilterTests.cs]: Fixed: https://github.com/dotliquid/dotliquid/pull/453/commits/80e1366110c8a64f680dd1cccfd2c344dd10effc

File: src/DotLiquid/Strainer.cs (hunk 15 above)
Comment (BenJBailey) [reply src/DotLiquid/Strainer.cs]: Making these explicitly public helped me realize they need to be in their own files. I moved them to the Util namespace: https://github.com/dotliquid/dotliquid/pull/453/commits/89f65da331b408202dcae2200262a6eb670146f9

File: src/DotLiquid/Strainer.cs (hunk 17 above)
Comment (BenJBailey) [reply src/DotLiquid/Strainer.cs]: Fixed: https://github.com/dotliquid/dotliquid/pull/453/commits/72a2410fe218c54f5da077e7012d54d35c87b34c

File: src/DotLiquid/Strainer.cs (hunk 19 above)
Comment (BenJBailey) [reply src/DotLiquid/Strainer.cs]: I fixed this here: https://github.com/dotliquid/dotliquid/pull/453/commits/72a2410fe218c54f5da077e7012d54d35c87b34c . I failed to make it's own commit.

File: src/DotLiquid/Strainer.cs (hunk 22 above)
Comment (BenJBailey) [reply src/DotLiquid/Strainer.cs]: Fixed: https://github.com/dotliquid/dotliquid/pull/453/commits/09d0be13de6eac9119cb701eccb5c37ad36fa46f

File: src/DotLiquid/Strainer.cs (hunk 23 above)
Comment (BenJBailey) [reply src/DotLiquid/Strainer.cs]: fixed: https://github.com/dotliquid/dotliquid/pull/453/commits/5026067905626c0bb3cc9e32348a9c6ebd9a270f

File: src/DotLiquid.Tests/FilterTests.cs (hunk 24 above)
Comment (BenJBailey) [reply src/DotLiquid.Tests/FilterTests.cs:356]: updated: https://github.com/dotliquid/dotliquid/pull/453/commits/fc59073b69e63abae0e66b343c938f05d0d15e30

**Comment by BenJBailey** [comment]
@daviburg I should have addressed all your concerns. Please let me know if I missed anything.

File: src/DotLiquid.Tests/FilterTests.cs [hunk 25]
```diff
@@ -360,18 +356,16 @@ public void TestFilterInContextOverridesMethodWithSameMethodSignaturesDifferentC
...
         [Test]
         public void TestFilterAsLocalOverridesMethodWithSameMethodSignaturesDifferentClasses()
         {
             Helper.AssertTemplateResult(
-                           expected: "ABClass Two",
+                           expected: "ABClass One",
                            template: "{{'A' | concatenate : 'B'}}",
                            localVariables: null,
-                           localFilters: new[] { typeof(FilterWithSameMethodSignatureDifferentClassOne), typeof(FilterWithSameMethodSignatureDifferentClassTwo) });
+                           localFilters: new[] {  typeof(FilterWithSameMethodSignatureDifferentClassTwo), typeof(FilterWithSameMethodSignatureDifferentClassOne) });
```
Comment (microalps) [inline src/DotLiquid.Tests/FilterTests.cs:366]: nit: @daviburg won't like this change to two spaces.

File: src/DotLiquid/Strainer.cs [hunk 26]
```diff
@@ -87,15 +57,14 @@ public Strainer(Context context)
...
-
         /// <summary>
         /// In this C# implementation, we can't use mixins. So we grab all the static
         /// methods from the specified type and use them instead.
         /// </summary>
         /// <param name="type"></param>
         public void Extend(Type type)
         {
-            // From what I can tell, calls to Extend should replace existing filters with the same number of params. So be it.
+            // Calls to Extend should replace existing filters with the same number of params.
```
Comment (microalps) [inline src/DotLiquid/Strainer.cs:67]: nit: remove "should" for the same reason.

**Review by microalps — CHANGES_REQUESTED** [review]

File: src/DotLiquid.Tests/FilterTests.cs (hunk 25 above)
Comment (BenJBailey) [reply src/DotLiquid.Tests/FilterTests.cs]: Fixed: https://github.com/dotliquid/dotliquid/pull/453/commits/10a524c01c5287c9ddfd6a794e12a8371917e549 
I'm sure @daviburg will appreciate it 😸

File: src/DotLiquid/Strainer.cs (hunk 26 above)
Comment (BenJBailey) [reply src/DotLiquid/Strainer.cs]: updated: https://github.com/dotliquid/dotliquid/pull/453/commits/faa0f59f2b6bd758fd666140edd2d75842a9004c

**Review by microalps — APPROVED** [review]

**Comment by microalps** [comment]
@daviburg please review and merge these if OK. I know there is an extra enter in his extension method, but let's not be too hard on first-time contributors!

**Review by daviburg — APPROVED** [review]
